core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,295 @@
+        AudioManager.AUDIOFOCUS_REQUEST_GRANTED
+    }
+  }
+
+  fun pauseOrResume() {
+    currentTTSTask?.let {
+      if (it.paused) {
+        if (!requestAudioFocus()) return@pauseOrResume
+      } else {
+        it.pause()
+      }
+    }
+  }
+
+  fun initWebView(webView: WebView) {
+    webView.addJavascriptInterface(TTSJavaScriptInterface(), "tts")
+  }
+
+  /**
+   * Releases the resources used by the engine.
+   *
+   * @see android.speech.tts.TextToSpeech.shutdown
+   */
+  fun shutdown() {
+    tts.shutdown()
+  }
+
+  /**
+   * The listener which is notified when initialization of the TextToSpeech engine is successfully
+   * done.
+   */
+  internal interface OnInitSucceedListener {
+    fun onInitSucceed()
+  }
+
+  /**
+   * The listener that is notified when speaking starts or stops (regardless of whether it was a
+   * result of error, user, or because whole text was read).
+   *
+   *
+   * Note that the methods of this interface may not be called from the UI thread.
+   */
+  interface OnSpeakingListener {
+    fun onSpeakingStarted()
+    fun onSpeakingEnded()
+  }
+
+  inner class TTSTask
+    (private val pieces: List<String>) {
+    private val currentPiece =
+      AtomicInteger(0)
+    @JvmField var paused = true
+    fun pause() {
+      paused = true
+      currentPiece.decrementAndGet()
+      tts.setOnUtteranceProgressListener(null)
+      tts.stop()
+    }
+
+    fun start() {
+      paused = (if (!paused) {
+        return
+      } else false)
+      val params = HashMap<String, String>()
+      // The utterance ID isn't actually used anywhere, the param is passed only to force
+      // the utterance listener to be notified
+      params[Engine.KEY_PARAM_UTTERANCE_ID] = "kiwixLastMessage"
+      if (currentPiece.get() < pieces.size) {
+        tts.speak(pieces[currentPiece.getAndIncrement()], QUEUE_ADD, params)
+      } else {
+        stop()
+      }
+      tts.setOnUtteranceProgressListener(object : UtteranceProgressListener() {
+        @SuppressWarnings("EmptyFunctionBlock")
+        override fun onStart(s: String) {
+        }
+
+        override fun onDone(s: String) {
+          val line: Int = currentPiece.toInt()
+          if (line >= pieces.size && !paused) {
+            stop()
+            return
+          }
+          tts.speak(pieces[line], QUEUE_ADD, params)
+          currentPiece.getAndIncrement()
+        }
+
+        override fun onError(s: String) {
+          Log.e(TAG_KIWIX, "TextToSpeech Error: $s")
+          context.toast(R.string.texttospeech_error, Toast.LENGTH_SHORT)
+        }
+      })
+    }
+
+    fun stop() {
+      currentTTSTask = null
+      onSpeakingListener.onSpeakingEnded()
+    }
+  }
+
+  private inner class TTSJavaScriptInterface {
+    @JavascriptInterface fun speakAloud(content: String) {
+      val splitted = content.split("[\\n.;]".toRegex()).toTypedArray()
+      val pieces = splitted.filter(String::isNotBlank)
```

Comment:
you also need to map them so they are trimmed to keep this equivalent to the previous code. You can also get rid of `.toTypedArray` and get rid of the `splitted` variable entirely by inlining it just keep each subsequent operation on a new line ie

```
val pieces= content.split(...)
                                .filter(...)
                                .map(....)
```

Reply:
@macgills In the filter method, it checks the string is not blank or not. Should we need map method for trim?